src/Npgsql.NodaTime/Internal/DateMultirangeHandler.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using NodaTime;
+using Npgsql.BackendMessages;
+using Npgsql.Internal;
+using Npgsql.Internal.TypeHandlers;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using NpgsqlTypes;
+
+namespace Npgsql.NodaTime.Internal
+{
+    public partial class DateMultirangeHandler : MultirangeHandler<LocalDate>,
+        INpgsqlTypeHandler<DateInterval[]>, INpgsqlTypeHandler<List<DateInterval>>
+    {
+        public DateMultirangeHandler(PostgresMultirangeType multirangePostgresType, NpgsqlTypeHandler subtypeHandler)
+            : base(multirangePostgresType, new DateRangeHandler(multirangePostgresType.Subrange, subtypeHandler))
+        {
+        }
+
+        public override Type GetFieldType(FieldDescription? fieldDescription = null) => typeof(DateInterval[]);
+        public override Type GetProviderSpecificFieldType(FieldDescription? fieldDescription = null) => typeof(DateInterval[]);
+
+        public override async ValueTask<object> ReadAsObject(NpgsqlReadBuffer buf, int len, bool async,
+            FieldDescription? fieldDescription = null)
+            => (await Read<DateInterval[]>(buf, len, async, fieldDescription))!;
+
+        async ValueTask<DateInterval[]> INpgsqlTypeHandler<DateInterval[]>.Read(
+            NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription)
+        {
+            await buf.Ensure(4, async);
+            var numRanges = buf.ReadInt32();
+            var multirange = new DateInterval[numRanges];
+
+            for (var i = 0; i < numRanges; i++)
```

Comment:
```suggestion
            for (var i = 0; i < multirange.Length; i++)
```

Allows jit to remove bounds check.